### PR TITLE
[Merged by Bors] - feat(Algebra/Group/Opposite): add `isLeft/RightCancelMul`

### PR DESCRIPTION
**Pull request by adomani**
[Zulip discussion](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/MulOpposites.20and.20.5Binstances.5D)

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Group/Opposite.lean [hunk 1]
```diff
@@ -102,11 +102,19 @@ instance leftCancelSemigroup [RightCancelSemigroup α] : LeftCancelSemigroup α
   { MulOpposite.semigroup α with
     mul_left_cancel := fun _ _ _ H => unop_injective <| mul_right_cancel <| op_injective H }
 
+@[to_additive]
+instance isLeftCancelSemigroup [Mul α] [IsRightCancelMul α] : IsLeftCancelMul αᵐᵒᵖ :=
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/Opposite.lean:106]:
```suggestion
instance isLeftCancelMul [Mul α] [IsRightCancelMul α] : IsLeftCancelMul αᵐᵒᵖ :=
```

File: Mathlib/Algebra/Group/Opposite.lean [hunk 2]
```diff
@@ -102,11 +102,19 @@ instance leftCancelSemigroup [RightCancelSemigroup α] : LeftCancelSemigroup α
   { MulOpposite.semigroup α with
     mul_left_cancel := fun _ _ _ H => unop_injective <| mul_right_cancel <| op_injective H }
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/Opposite.lean:103]: If you swap the order of the declarations, you should be able to prove this one in terms of the new one.

Comment (adomani) [reply Mathlib/Algebra/Group/Opposite.lean]: Your suggestion worked.  Several of the `MulOpposite.[something] α with` could be removed and Lean would not complain.  Is there a reason to keep them?  I have been following as a spectator the "`let` in instances" progress, but I do not know if this is part of it.

File: Mathlib/Algebra/Group/Opposite.lean [hunk 3]
```diff
@@ -92,6 +92,14 @@ We also generate additive structures on `αᵃᵒᵖ` using `to_additive`
 -/
 
 
+@[to_additive]
+instance isRightCancelMul [Mul α] [IsLeftCancelMul α] : IsRightCancelMul αᵐᵒᵖ :=
+  { mul_right_cancel := fun _ _ _ => (unop_injective <| mul_left_cancel <| op_injective ·) }
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/Opposite.lean:97]: a bit tidier using the new syntax:
```suggestion
instance isRightCancelMul [Mul α] [IsLeftCancelMul α] : IsRightCancelMul αᵐᵒᵖ where
  mul_right_cancel _ _ _ := (unop_injective <| mul_left_cancel <| op_injective ·)
```

File: Mathlib/Algebra/Group/Opposite.lean [hunk 4]
```diff
@@ -100,12 +108,12 @@ instance semigroup [Semigroup α] : Semigroup αᵐᵒᵖ :=
...
 instance leftCancelSemigroup [RightCancelSemigroup α] : LeftCancelSemigroup αᵐᵒᵖ :=
   { MulOpposite.semigroup α with
-    mul_left_cancel := fun _ _ _ H => unop_injective <| mul_right_cancel <| op_injective H }
+    mul_left_cancel := by apply mul_left_cancel }
 
 @[to_additive]
 instance rightCancelSemigroup [LeftCancelSemigroup α] : RightCancelSemigroup αᵐᵒᵖ :=
   { MulOpposite.semigroup α with
-    mul_right_cancel := fun _ _ _ H => unop_injective <| mul_left_cancel <| op_injective H }
+    mul_right_cancel := by apply mul_right_cancel }
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/Opposite.lean:116]: if you switch this to `where`, you can write
```suggestion
    mul_right_cancel _ _ := mul_right_cancel
```
or similar.

**Review by eric-wieser — APPROVED** [review]
bors d+

Thanks!

File: Mathlib/Algebra/Group/Opposite.lean [hunk 5]
```diff
@@ -93,24 +93,28 @@ We also generate additive structures on `αᵃᵒᵖ` using `to_additive`
 
 
 @[to_additive]
-instance semigroup [Semigroup α] : Semigroup αᵐᵒᵖ :=
-  { MulOpposite.mul α with
-    mul_assoc := fun x y z => unop_injective <| Eq.symm <| mul_assoc (unop z) (unop y) (unop x) }
+instance isRightCancelMul [Mul α] [IsLeftCancelMul α] : IsRightCancelMul αᵐᵒᵖ where
+  mul_right_cancel := fun _ _ _ => (unop_injective <| mul_left_cancel <| op_injective ·)
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/Opposite.lean:97]: The point in `where` is that you can write
```suggestion
  mul_right_cancel _ _ _ := (unop_injective <| mul_left_cancel <| op_injective ·)
```

Comment (adomani) [reply Mathlib/Algebra/Group/Opposite.lean]: Yes, I only noticed it later...  I force pushed a version without the arrow!

**Pull request closed without merging**